operators/test/e2e/cmd/run/run.go

```diff
@@ -0,0 +1,404 @@
+func (s *helper) monitorTestJob(client *kubernetes.Clientset) error {
+	log.Info("Waiting for test job to start")
+	ctx, cancelFunc := context.WithTimeout(context.Background(), jobTimeout)
+	defer func() {
+		cancelFunc()
+		runtime.HandleCrash()
+	}()
+
+	factory := informers.NewSharedInformerFactoryWithOptions(client, kubePollInterval,
+		informers.WithNamespace(s.testContext.E2ENamespace),
+		informers.WithTweakListOptions(func(opt *metav1.ListOptions) {
+			opt.LabelSelector = fmt.Sprintf("%s=%s", testRunLabel, s.testContext.TestRun)
+		}))
+
+	informer := factory.Core().V1().Pods().Informer()
+
+	jobStarted := false
+	streamStatus := make(chan error, 1)
+	var err error
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*corev1.Pod)
+			log.Info("Pod added", "name", pod.Name)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newPod := newObj.(*corev1.Pod)
+			switch newPod.Status.Phase {
+			case corev1.PodRunning:
+				if !jobStarted {
+					jobStarted = true
+					log.Info("Pod started", "name", newPod.Name)
+					go s.streamTestJobOutput(streamStatus, client, newPod.Name)
+				} else {
+					select {
+					case streamErr := <-streamStatus:
+						if streamErr != nil {
+							log.Error(streamErr, "Stream failure")
+							err = streamErr
+							cancelFunc()
+						}
+					default:
+					}
+				}
+			case corev1.PodSucceeded:
+				log.Info("Tests completed successfully", "name", newPod.Name)
+				cancelFunc()
+			case corev1.PodFailed:
+				log.Info("Pod is in failed state", "name", newPod.Name)
+				err = errors.New("tests failed")
+				cancelFunc()
+			default:
+				log.Info("Waiting for pod to be ready", "name", newPod.Name, "status", newPod.Status.Phase)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*corev1.Pod)
+			log.Info("Pod deleted", "name", pod.Name)
+			cancelFunc()
+		},
+	})
+
+	informer.Run(ctx.Done())
+	return err
+}
+
+func (s *helper) streamTestJobOutput(streamStatus chan<- error, client *kubernetes.Clientset, pod string) {
+	log.Info("Streaming pod logs", "name", pod)
+	defer close(streamStatus)
+	sinceSeconds := int64(30)
+	opts := &corev1.PodLogOptions{
+		Container:    "e2e", //TODO remove hardcoded value
+		Follow:       true,
+		SinceSeconds: &sinceSeconds,
+	}
+
+	req := client.CoreV1().Pods(s.testContext.E2ENamespace).GetLogs(pod, opts).Context(context.Background())
+	stream, err := req.Stream()
+	if err != nil {
+		streamStatus <- err
+		return
+	}
+	defer stream.Close()
+
+	var buffer [logBufferSize]byte
+	if _, err := io.CopyBuffer(os.Stdout, stream, buffer[:]); err != nil {
+		if err == io.EOF {
+			log.Info("Log stream ended")
+			return
+		}
+
+		streamStatus <- err
+	}
+}
+
+func (s *helper) kubectlApplyTemplate(templatePath string, templateParam interface{}, deleteOnExit bool) {
+	if s.err != nil {
+		return
+	}
+
+	outFilePath, err := s.renderTemplate(templatePath, templateParam)
+	if err != nil {
+		return
```

Comment:
It's already captured by the error field in the helper.

Reply:
Hm, I missed that. But does that not raise the possibility of swallowing errors due to programmer error e.g. 
* when you overwrite accidentally  a previous error recorded there. It would be quite tricky to figure out in the code that that never happens e.g. for a reviewer. 
* when you forget to 'store' the error in your struct

Is there a reason not to bubble up the errors via return types?

If need be you can wrap some kind of step abstraction around it (which you already started by putting the step into a slice)?

Reply:
Initially I tried to follow the pattern described [here](https://blog.golang.org/errors-are-values) to avoid repetitive error checks. However, you're right, it does force the programmers to think more carefully about errors. I'll see if it can be switched back to using normal error return values.